app/routes.py

```diff
@@ -87,10 +89,32 @@ def logout():
 # =========================
 # 2. Login Required
 # # =========================
+@app.route('/bookmark/<book_id>', methods=['GET'])
+@login_required
+def bookmark(book_id):
+    bookmark = Bookmark(user_id=current_user.get_id(), book_id=int(book_id))

```

Comment:
Before creating a new bookmark, check whether this user has already bookmarked the book, by checking

```
if  Bookmark.query.filter_by(user_id=current_user.get_id(), book_id=int(book_id)).first() is None:
    Bookmark(user_id=current_user.get_id(), book_id=int(book_id))
```